theseus/labs/lie/lie_tensor.py

```diff
@@ -32,35 +57,115 @@ def _get_fn_lib(ltype: _ltype):
     }[ltype]
 
 
-class LieTensor:
-    def __init__(self, data_tensor: torch.Tensor, ltype: _ltype):
-        self._t = data_tensor
-        self._fn_lib = _get_fn_lib(ltype)
-        self._fn_lib.check_group_tensor(data_tensor)
+class _LieTensorBase(torch.Tensor):
+    def __new__(cls, data: Any, ltype: _ltype, requires_grad=None):
+        if requires_grad is None:
+            return super().__new__(cls, data)  # type: ignore
+        else:
+            return cls._make_subclass(cls, data, requires_grad)  # type: ignore
+
+    @property
+    def _t(self) -> torch.Tensor:
+        return self.as_subclass(torch.Tensor)
+
+    def __init__(self, data: Any, ltype: _ltype, requires_grad=None):
         self.ltype = ltype
 
-    def __repr__(self) -> str:
+    def __repr__(self) -> str:  # type: ignore
         return f"LieTensor({self._t}, ltype=lie.{self.ltype})"
 
+    @classmethod
+    def to_torch(cls, args: Any):
+        def _resolve(x):
+            return x._t if isinstance(x, cls) else x
+
+        return tree_map(_resolve, args)
+
+    @classmethod
+    def resolve_ltype(cls, t: Any):
+        return t.ltype if isinstance(t, cls) else None
+
+    @classmethod
+    def _torch_func_impl_eucl(cls, func, types, args=(), kwargs=None, raw_tensor=False):
+        kwargs = kwargs or {}
+        torch_args = cls.to_torch(args)
+        torch_kwargs = cls.to_torch(kwargs)
+        ltypes = [cls.resolve_ltype(a) for a in tree_flatten(args)[0]]
+        ltypes = set(x for x in ltypes if x is not None)
+        if len(ltypes) > 1:
+            raise ValueError(
+                f"All LieTensors must be of the same ltype. " f"But found {ltypes}"
+            )
+        ltype = next(iter(ltypes)) if len(ltypes) > 0 else None
+        ret = func(*torch_args, **torch_kwargs)
+        return (
+            cls(ret, ltype) if isinstance(ret, torch.Tensor) and not raw_tensor else ret
```

Comment:
nit: you most likely want a tree_map_only(torch.Tensor, lambda t: cls(t, ltype), ret)` here to cover functions returning multiple things.